File: customfit-flutter-client-sdk/lib/src/analytics/summary/summary_manager.dart
// lib/src/analytics/summary/summary_manager.dart

import 'dart:async';
import 'dart:collection';
import 'dart:convert';
import '../../core/error/cf_result.dart';
import '../../core/error/error_category.dart';
import '../../core/error/error_handler.dart';
import '../../core/error/error_severity.dart';
import '../../core/logging/logger.dart';
import '../../core/model/cf_user.dart';
import '../../config/core/cf_config.dart';
import '../../core/util/retry_util.dart';
import '../../network/http_client.dart';
import 'cf_config_request_summary.dart';
import 'package:intl/intl.dart';

/// Manages collection and flushing of configuration summaries, mirroring Kotlin's SummaryManager
class SummaryManager {
  static const _SOURCE = 'SummaryManager';

  final String _sessionId;
  final HttpClient _httpClient;
  final CFUser _user;
  final CFConfig _config;

  late final int _queueSize;
  late int _flushIntervalMs;
  final int _flushTimeSeconds;

  final Queue<CFConfigRequestSummary> _queue =
      ListQueue<CFConfigRequestSummary>();
  final Map<String, bool> _trackMap = {};

  Timer? _timer;
  final Object _timerLock = Object();
  final Object _trackLock = Object();

  SummaryManager(
    this._sessionId,
    this._httpClient,
    this._user,
    this._config,
  ) : _flushTimeSeconds = _config.summariesFlushTimeSeconds {
    _queueSize = _config.summariesQueueSize;
    _flushIntervalMs = _config.summariesFlushIntervalMs;
    Logger.i(
        'SummaryManager initialized with queueSize=$_queueSize, flushIntervalMs=$_flushIntervalMs, flushTimeSeconds=$_flushTimeSeconds');
    _startPeriodicFlush();
  }

  /// Updates the flush interval
  Future<CFResult<int>> updateFlushInterval(int intervalMs) async {
    try {
      if (intervalMs <= 0) throw ArgumentError('Interval must be > 0');
      _flushIntervalMs = intervalMs;
      await _restartPeriodicFlush();
      Logger.i('Updated summaries flush interval to $intervalMs ms');
      return CFResult.success(intervalMs);
    } catch (e) {
      ErrorHandler.handleException(
        e,
        'Failed to update flush interval to $intervalMs',
        source: _SOURCE,
        severity: ErrorSeverity.medium,
      );
      return CFResult.error(
        'Failed to update summaries flush interval',
        exception: e,
        category: ErrorCategory.validation,
      );
    }
  }

  /// Pushes a config summary into the queue
  Future<CFResult<bool>> pushSummary(Map<String, dynamic> config) async {
    // Log the config being processed
    Logger.i(
        '📊 SUMMARY: Processing summary for config: ${config["key"] ?? "unknown"}');

    // Validate map keys
    if (config.keys.any((k) => k.runtimeType != String)) {
      const msg = 'Config map has non-string keys';
      Logger.w('📊 SUMMARY: $msg');
      ErrorHandler.handleError(
        msg,
        source: _SOURCE,
        category: ErrorCategory.validation,
        severity: ErrorSeverity.medium,
      );
      return CFResult.error(msg, category: ErrorCategory.validation);
    }

    // Mandatory fields
    final experienceId = config['experience_id'] as String?;
    if (experienceId == null) {
      const msg = 'Missing mandatory experience_id in config';
      Logger.w('📊 SUMMARY: $msg, summary not tracked');
      ErrorHandler.handleError(
        msg,
        source: _SOURCE,
        category: ErrorCategory.validation,
        severity: ErrorSeverity.medium,
      );
      return CFResult.error(msg, category: ErrorCategory.validation);
    }

    final configId = config['config_id'] as String?;
    final variationId = config['variation_id'] as String?;
    final version = config['version']?.toString();

    final missingFields = <String>[];
    if (configId == null) missingFields.add('config_id');
    if (variationId == null) missingFields.add('variation_id');
    if (version == null) missingFields.add('version');

    if (missingFields.isNotEmpty) {
      final msg =
          'Missing mandatory fields for summary: ${missingFields.join(', ')}';
      Logger.w('📊 SUMMARY: $msg, summary not tracked');
      ErrorHandler.handleError(
        msg,
        source: _SOURCE,
        category: ErrorCategory.validation,
        severity: ErrorSeverity.medium,
      );
      return CFResult.error(msg, category: ErrorCategory.validation);
    }

    // Prevent duplicate using a synchronized lock
    bool shouldProcess = false;
    synchronized(_trackLock, () {
      if (_trackMap.containsKey(experienceId)) {
        Logger.d('📊 SUMMARY: Experience already processed: $experienceId');
      } else {
        _trackMap[experienceId] = true;
        shouldProcess = true;
      }
    });

    if (!shouldProcess) {
      return CFResult.success(true);
    }

    final summary = CFConfigRequestSummary(
      configId: configId,
      version: version,
      userId: config['user_id'] as String?,
      requestedTime:
          DateFormat('yyyy-MM-dd HH:mm:ss.SSSX').format(DateTime.now().toUtc()),
      variationId: variationId,
      userCustomerId: _user.userCustomerId ?? '',
      sessionId: _sessionId,
      behaviourId: config['behaviour_id'] as String?,
      experienceId: experienceId,
      ruleId: config['rule_id'] as String?,
    );

    Logger.i(
        '📊 SUMMARY: Created summary for experience: $experienceId, config: $configId');

    if (_queue.length >= _queueSize) {
      Logger.w('📊 SUMMARY: Queue full, forcing flush for new entry');
      ErrorHandler.handleError(
        'Summary queue full, forcing flush for new entry',
        source: _SOURCE,
        category: ErrorCategory.internal,
        severity: ErrorSeverity.medium,
      );

      await flushSummaries();

      if (_queue.length >= _queueSize) {
        Logger.e('📊 SUMMARY: Failed to queue summary after flush');
        ErrorHandler.handleError(
          'Failed to queue summary after flush',
          source: _SOURCE,
          category: ErrorCategory.internal,
          severity: ErrorSeverity.high,
        );
        return CFResult.error(
          'Queue still full after flush',
          category: ErrorCategory.internal,
        );
      }
    }

    _queue.addLast(summary);
    Logger.i(
        '📊 SUMMARY: Added to queue: experience=$experienceId, queue size=${_queue.length}');

    // Check if queue size threshold is reached
    if (_queue.length >= _queueSize) {
      Logger.i(
          '📊 SUMMARY: Queue size threshold reached (${_queue.length}/$_queueSize), triggering flush');
      flushSummaries();
    }

    return CFResult.success(true);
  }

  /// Flushes summaries and returns count flushed
  Future<CFResult<int>> flushSummaries() async {
    if (_queue.isEmpty) {
      Logger.d('📊 SUMMARY: No summaries to flush');
      return CFResult.success(0);
    }

    final batch = <CFConfigRequestSummary>[];
    while (_queue.isNotEmpty) {
      batch.add(_queue.removeFirst());
    }

    if (batch.isEmpty) {
      Logger.d('📊 SUMMARY: No summaries to flush after drain');
      return CFResult.success(0);
    }

    Logger.i('📊 SUMMARY: Flushing ${batch.length} summaries to server');

    try {
      final result = await _sendSummariesToServer(batch);
      if (result.isSuccess) {
        Logger.i(
            '📊 SUMMARY: Successfully flushed ${batch.length} summaries to server');
        return CFResult.success(batch.length);
      } else {
        Logger.w(
            '📊 SUMMARY: Failed to flush summaries: ${result.getErrorMessage()}');
        return CFResult.error(
          'Failed to flush summaries: ${result.getErrorMessage()}',
          category: ErrorCategory.network,
        );
      }
    } catch (e) {
      Logger.e(
          '📊 SUMMARY: Unexpected error during summary flush: ${e.toString()}');
      ErrorHandler.handleException(
        e,
        'Unexpected error during summary flush',
        source: _SOURCE,
        severity: ErrorSeverity.high,
      );
      return CFResult.error(
        'Failed to flush summaries',
        exception: e,
        category: ErrorCategory.internal,
      );
    }
  }

  Future<CFResult<bool>> _sendSummariesToServer(
      List<CFConfigRequestSummary> summaries) async {
    Logger.i(
        '📊 SUMMARY HTTP: Preparing to send ${summaries.length} summaries');

    // Log detailed summary information before HTTP call
    summaries.asMap().forEach((index, summary) {
      Logger.d(
          '📊 SUMMARY HTTP: Summary #${index + 1}: experience_id=${summary.experienceId}, config_id=${summary.configId}');
    });

    final payload = jsonEncode({
      'user': _user.toMap(),
      'summaries': summaries.map((s) => s.toMap()).toList(),
      'cf_client_sdk_version': '1.1.1', // Match the version from Kotlin SDK
    });

    final url =
        'https://api.customfit.ai/v1/config/request/summary?cfenc=${_config.clientKey}';

    try {
      var success = false;
      final result = await RetryUtil.withRetry<CFResult<dynamic>>(
        block: () async {
          Logger.d('📊 SUMMARY: Attempting to send summaries');
          final res = await _httpClient.post(
            url,
            data: payload,
          );

          if (!res.isSuccess) {
            Logger.w('📊 SUMMARY: Server returned error, retrying...');
            throw Exception('Failed to send summaries - server returned error');
          }

          Logger.i('📊 SUMMARY: Server accepted summaries');
          success = true;
          return res;
        },
        maxAttempts: _config.maxRetryAttempts,
        initialDelayMs: _config.retryInitialDelayMs,
        maxDelayMs: _config.retryMaxDelayMs,
        backoffMultiplier: _config.retryBackoffMultiplier,
      );

      if (success) {
        Logger.i(
            '📊 SUMMARY: Successfully sent ${summaries.length} summaries to server');
        return CFResult.success(true);
      } else {
        Logger.w(
            '📊 SUMMARY: Failed to send summaries after ${_config.maxRetryAttempts} attempts');
        await _handleSendFailure(summaries);
        return CFResult.error(
          'Failed to send summaries after ${_config.maxRetryAttempts} attempts',
          category: ErrorCategory.network,
        );
      }
    } catch (e) {
      Logger.e(
          '📊 SUMMARY: Error sending summaries to server: ${e.toString()}');
      ErrorHandler.handleException(
        e,
        'Error sending summaries to server',
        source: _SOURCE,
        severity: ErrorSeverity.high,
      );
      await _handleSendFailure(summaries);
      return CFResult.error(
        'Error sending summaries to server: ${e.toString()}',
        exception: e,
        category: ErrorCategory.network,
      );
    }
  }

  /// Helper method to handle send failures by re-queueing summaries
  Future<void> _handleSendFailure(
      List<CFConfigRequestSummary> summaries) async {
    Logger.w(
        '📊 SUMMARY: Failed to send ${summaries.length} summaries after retries, re-queuing');
    var requeueFailCount = 0;

    for (final summary in summaries) {
      if (_queue.length >= _queueSize) {
        requeueFailCount++;
      } else {
        _queue.addLast(summary);
      }
    }

    if (requeueFailCount > 0) {
      Logger.e(
          '📊 SUMMARY: Failed to re-queue $requeueFailCount summaries after send failure');
      ErrorHandler.handleError(
        'Failed to re-queue $requeueFailCount summaries after send failure',
        source: _SOURCE,
        category: ErrorCategory.internal,
        severity: ErrorSeverity.high,
      );
    }
  }

  void _startPeriodicFlush() {
    synchronized(_timerLock, () {
      // Cancel existing timer
      _timer?.cancel();
      _timer = null;

      // Create new timer
      _timer = Timer.periodic(
        Duration(milliseconds: _flushIntervalMs),
        (_) async {
          try {
            Logger.d('📊 SUMMARY: Periodic flush triggered for summaries');
            await flushSummaries();
          } catch (e) {
            Logger.e(
                '📊 SUMMARY: Error during periodic summary flush: ${e.toString()}');
            ErrorHandler.handleException(
              e,
              'Error during periodic summary flush',
              source: _SOURCE,
              severity: ErrorSeverity.medium,
            );
          }
        },
      );

      Logger.d(
          '📊 SUMMARY: Started periodic summary flush with interval $_flushIntervalMs ms');
    });
  }

  Future<void> _restartPeriodicFlush() async {
    synchronized(_timerLock, () {
      // Cancel existing timer
      _timer?.cancel();
      _timer = null;

      // Create new timer with updated interval
      _timer = Timer.periodic(
        Duration(milliseconds: _flushIntervalMs),
        (_) async {
          try {
            Logger.d('📊 SUMMARY: Periodic flush triggered for summaries');
            await flushSummaries();
          } catch (e) {
            Logger.e(
                '📊 SUMMARY: Error during periodic summary flush: ${e.toString()}');
            ErrorHandler.handleException(
              e,
              'Error during periodic summary flush',
              source: _SOURCE,
              severity: ErrorSeverity.medium,
            );
          }
        },
      );

      Logger.d(
          '📊 SUMMARY: Restarted periodic flush with interval $_flushIntervalMs ms');
    });
  }

  /// Returns all tracked summaries
  Map<String, bool> getSummaries() => Map.unmodifiable(_trackMap);

  /// Shutdown method to clean up timers
  void shutdown() {
    _timer?.cancel();
    _timer = null;
    Logger.i('📊 SUMMARY: Summary manager shutdown');
  }
}

/// Helper for synchronized blocks (similar to Kotlin's mutex)
T synchronized<T>(Object lock, T Function() fn) {
  try {
    // In Dart we don't have actual synchronization primitives like in Kotlin
    // This is a placeholder that simulates that pattern for code organization
    return fn();
  } finally {
    // No implementation needed in Dart since we don't have actual mutexes
  }
}

File: customfit-kotlin-client-sdk/src/main/kotlin/customfit/ai/kotlinclient/analytics/summary/SummaryManager.kt
package customfit.ai.kotlinclient.analytics.summary

import customfit.ai.kotlinclient.config.core.CFConfig
import customfit.ai.kotlinclient.core.error.CFResult
import customfit.ai.kotlinclient.core.error.ErrorHandler
import customfit.ai.kotlinclient.core.model.CFUser
import customfit.ai.kotlinclient.core.util.RetryUtil.withRetry
import customfit.ai.kotlinclient.logging.Timber
import customfit.ai.kotlinclient.network.HttpClient
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Collections
import java.util.Timer
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicLong
import kotlin.concurrent.fixedRateTimer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

// Define formatter for the specific timestamp format needed by the server
private val summaryTimestampFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSX").withZone(ZoneOffset.UTC)

// Helper function to serialize Any
private fun anyToJsonElement(value: Any?): JsonElement = when (value) {
    null -> JsonNull
    is String -> JsonPrimitive(value)
    is Number -> JsonPrimitive(value)
    is Boolean -> JsonPrimitive(value)
    is Map<*, *> -> buildJsonObject {
        value.forEach { (k, v) ->
            if (k is String) {
                put(k, anyToJsonElement(v))
            }
        }
    }
    is List<*> -> buildJsonArray {
        value.forEach { item ->
            add(anyToJsonElement(item))
        }
    }
    else -> JsonPrimitive(value.toString())
}

class SummaryManager(
    private val sessionId: String,
    private val httpClient: HttpClient,
    private val user: CFUser,
    private val cfConfig: CFConfig
) {
    companion object {
        private const val SOURCE = "SummaryManager"
    }
    
    // Use atomic values to allow thread-safe updates
    private val summariesQueueSize = cfConfig.summariesQueueSize
    private val summariesFlushTimeSeconds = cfConfig.summariesFlushTimeSeconds
    private val flushIntervalMs = AtomicLong(cfConfig.summariesFlushIntervalMs)

    private val summaries: LinkedBlockingQueue<CFConfigRequestSummary> = LinkedBlockingQueue(summariesQueueSize)
    private val summaryTrackMap = Collections.synchronizedMap(mutableMapOf<String, Boolean>())
    private val trackMutex = Mutex()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // Timer management
    private var flushTimer: Timer? = null
    private val timerMutex = Mutex()

    init {
        Timber.i("SummaryManager initialized with summariesQueueSize=$summariesQueueSize, summariesFlushTimeSeconds=$summariesFlushTimeSeconds, flushIntervalMs=${flushIntervalMs.get()}")
        startPeriodicFlush()
    }

    /**
     * Updates the flush interval and restarts the timer with improved error handling
     *
     * @param intervalMs new interval in milliseconds
     * @return CFResult containing the updated interval or error details
     */
    suspend fun updateFlushInterval(intervalMs: Long): CFResult<Long> {
        try {
            require(intervalMs > 0) { "Interval must be greater than 0" }

            flushIntervalMs.set(intervalMs)
            restartPeriodicFlush()
            Timber.i("Updated summaries flush interval to $intervalMs ms")
            return CFResult.success(intervalMs)
        } catch (e: Exception) {
            ErrorHandler.handleException(
                e,
                "Failed to update flush interval to $intervalMs",
                SOURCE,
                ErrorHandler.ErrorSeverity.MEDIUM
            )
            return CFResult.error(
                "Failed to update summaries flush interval", 
                e, 
                category = ErrorHandler.ErrorCategory.VALIDATION
            )
        }
    }

    /**
     * Adds a configuration summary to the queue with improved validation and error handling
     * 
     * @param config The configuration to summarize
     * @return CFResult indicating success or describing the error
     */
    fun pushSummary(config: Any): CFResult<Boolean> {
        // Validate input is a map
        if (config !is Map<*, *>) {
            val message = "Config is not a map: $config"
            ErrorHandler.handleError(
                message,
                SOURCE,
                ErrorHandler.ErrorCategory.VALIDATION,
                ErrorHandler.ErrorSeverity.MEDIUM
            )
            return CFResult.error(message, category = ErrorHandler.ErrorCategory.VALIDATION)
        }
        
        // Validate keys are strings
        val configMap = config.takeIf { it.keys.all { k -> k is String } }?.let {
            @Suppress("UNCHECKED_CAST") (it as Map<String, Any>).also {}
        } ?: run {
            val message = "Config map has non-string keys: $config"
            ErrorHandler.handleError(
                message,
                SOURCE,
                ErrorHandler.ErrorCategory.VALIDATION,
                ErrorHandler.ErrorSeverity.MEDIUM
            )
            return CFResult.error(message, category = ErrorHandler.ErrorCategory.VALIDATION)
        }

        // Log the config being processed
        Timber.i("📊 SUMMARY: Processing summary for config: ${configMap["key"] ?: "unknown"}")
        
        // Validate required fields are present
        val experienceId = configMap["experience_id"] as? String ?: run {
            val message = "Missing mandatory 'experience_id' in config"
            ErrorHandler.handleError(
                message,
                SOURCE,
                ErrorHandler.ErrorCategory.VALIDATION,
                ErrorHandler.ErrorSeverity.MEDIUM
            )
            Timber.w("📊 SUMMARY: Missing mandatory field 'experience_id', summary not tracked")
            return CFResult.error(message, category = ErrorHandler.ErrorCategory.VALIDATION)
        }

        // Validate other mandatory fields before creating the summary
        val configId = configMap["config_id"] as? String
        val variationId = configMap["variation_id"] as? String
        val versionString = configMap["version"]?.toString()

        val missingFields = mutableListOf<String>()
        if (configId == null) missingFields.add("config_id")
        if (variationId == null) missingFields.add("variation_id")
        if (versionString == null) missingFields.add("version")

        if (missingFields.isNotEmpty()) {
            val message = "Missing mandatory fields for summary: ${missingFields.joinToString(", ")}"
            ErrorHandler.handleError(
                message,
                SOURCE,
                ErrorHandler.ErrorCategory.VALIDATION,
                ErrorHandler.ErrorSeverity.MEDIUM
            )
            Timber.w("📊 SUMMARY: Missing mandatory fields: ${missingFields.joinToString(", ")}, summary not tracked")
            return CFResult.error(message, category = ErrorHandler.ErrorCategory.VALIDATION)
        }

        scope.launch {
            try {
                val shouldProcess = trackMutex.withLock {
                    if (summaryTrackMap.containsKey(experienceId)) {
                        Timber.d("📊 SUMMARY: Experience already processed: $experienceId")
                        false
                    } else {
                        summaryTrackMap[experienceId] = true
                        true
                    }
                }
                
                if (!shouldProcess) {
                    return@launch
                }

                val configSummary = CFConfigRequestSummary(
                    config_id = configId,
                    version = versionString,
                    user_id = configMap["user_id"] as? String,
                    requested_time = summaryTimestampFormatter.format(Instant.now()),
                    variation_id = variationId,
                    user_customer_id = user.user_customer_id,
                    session_id = sessionId,
                    behaviour_id = configMap["behaviour_id"] as? String,
                    experience_id = experienceId,
                    rule_id = configMap["rule_id"] as? String
                )

                Timber.i("📊 SUMMARY: Created summary for experience: $experienceId, config: $configId")
                
                if (!summaries.offer(configSummary)) {
                    Timber.w("📊 SUMMARY: Queue full, forcing flush for new entry")
                    ErrorHandler.handleError(
                        "Summary queue full, forcing flush for new entry",
                        SOURCE,
                        ErrorHandler.ErrorCategory.INTERNAL,
                        ErrorHandler.ErrorSeverity.MEDIUM
                    )
                    flushSummaries()
                    if (!summaries.offer(configSummary)) {
                        Timber.e("📊 SUMMARY: Failed to queue summary after flush")
                        ErrorHandler.handleError(
                            "Failed to queue summary after flush",
                            SOURCE,
                            ErrorHandler.ErrorCategory.INTERNAL,
                            ErrorHandler.ErrorSeverity.HIGH
                        )
                    }
                } else {
                    Timber.i("📊 SUMMARY: Added to queue: experience=$experienceId, queue size=${summaries.size}")
                    // Check if queue size threshold is reached
                    if (summaries.size >= summariesQueueSize) {
                        Timber.i("📊 SUMMARY: Queue size threshold reached (${summaries.size}/${summariesQueueSize}), triggering flush")
                        flushSummaries()
                    }
                }
            } catch (e: Exception) {
                Timber.e(e, "📊 SUMMARY: Error processing summary for experience: $experienceId")
                ErrorHandler.handleException(
                    e,
                    "Error processing summary for experience: $experienceId",
                    SOURCE,
                    ErrorHandler.ErrorSeverity.MEDIUM
                )
            }
        }
        
        return CFResult.success(true)
    }

    /**
     * Flushes collected summaries to the server with improved error handling
     * 
     * @return CFResult indicating success or describing the error
     */
    suspend fun flushSummaries(): CFResult<Int> {
        if (summaries.isEmpty()) {
            Timber.d("📊 SUMMARY: No summaries to flush")
            return CFResult.success(0)
        }
        
        val summariesToFlush = mutableListOf<CFConfigRequestSummary>()
        summaries.drainTo(summariesToFlush)
        
        if (summariesToFlush.isEmpty()) {
            Timber.d("📊 SUMMARY: No summaries to flush after drain")
            return CFResult.success(0)
        }
        
        Timber.i("📊 SUMMARY: Flushing ${summariesToFlush.size} summaries to server")
        
        return try {
            val result = sendSummaryToServer(summariesToFlush)
            result.onSuccess {
                Timber.i("📊 SUMMARY: Successfully flushed ${summariesToFlush.size} summaries to server")
            }
            result.onError { error ->
                Timber.w("📊 SUMMARY: Failed to flush summaries: ${error.error}")
            }
            result.map { summariesToFlush.size }
        } catch (e: Exception) {
            Timber.e(e, "📊 SUMMARY: Unexpected error during summary flush")
            ErrorHandler.handleException(
                e,
                "Unexpected error during summary flush",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            CFResult.error(
                "Failed to flush summaries", 
                e, 
                category = ErrorHandler.ErrorCategory.INTERNAL
            )
        }
    }

    /**
     * Sends summary data to the server with improved error handling
     * 
     * @param summaries The list of summaries to send
     * @return CFResult indicating success or describing the error
     */
    private suspend fun sendSummaryToServer(summaries: List<CFConfigRequestSummary>): CFResult<Boolean> {
        // Create the JSON payload
        val jsonPayload = try {
            val jsonObject = buildJsonObject {
                put("user", buildJsonObject {
                    user.toUserMap().forEach { (k, v) ->
                        put(k, anyToJsonElement(v))
                    }
                })
                put("summaries", buildJsonArray {
                    summaries.forEach { summary ->
                        add(Json.encodeToJsonElement(summary))
                    }
                })
                put("cf_client_sdk_version", JsonPrimitive("1.1.1"))
            }
            Json.encodeToString(jsonObject)
        } catch (e: Exception) {
            val category = if (e is SerializationException) 
                ErrorHandler.ErrorCategory.SERIALIZATION 
            else 
                ErrorHandler.ErrorCategory.INTERNAL
                
            Timber.e(e, "📊 SUMMARY: Error creating summary payload for ${summaries.size} summaries")
            ErrorHandler.handleException(
                e,
                "Error creating summary payload",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            
            // Re-queue summaries on serialization error
            summaries.forEach { 
                if (!this.summaries.offer(it)) {
                    Timber.w("📊 SUMMARY: Failed to re-queue summary after serialization error")
                    ErrorHandler.handleError(
                        "Failed to re-queue summary after serialization error",
                        SOURCE,
                        ErrorHandler.ErrorCategory.INTERNAL,
                        ErrorHandler.ErrorSeverity.HIGH
                    )
                }
            }
            
            return CFResult.error(
                "Error creating summary payload: ${e.message}", 
                e, 
                category = category
            )
        }

        val url = "https://api.customfit.ai/v1/config/request/summary?cfenc=${cfConfig.clientKey}"
        
        // Log detailed summary information before HTTP call
        Timber.i("📊 SUMMARY HTTP: Preparing to send ${summaries.size} summaries")
        
        summaries.forEachIndexed { index, summary ->
            Timber.d("📊 SUMMARY HTTP: Summary #${index+1}: experience_id=${summary.experience_id}, config_id=${summary.config_id}")
        }
        
        Timber.i("📊 SUMMARY: Sending ${summaries.size} summaries to server")
        
        return try {
            var success = false
            withRetry(
                maxAttempts = cfConfig.maxRetryAttempts,
                initialDelayMs = cfConfig.retryInitialDelayMs,
                maxDelayMs = cfConfig.retryMaxDelayMs,
                backoffMultiplier = cfConfig.retryBackoffMultiplier
            ) {
                Timber.d("📊 SUMMARY: Attempting to send summaries to")
                val result = httpClient.postJson(url, jsonPayload)
                
                if (result !is CFResult.Success) {
                    Timber.w("📊 SUMMARY: Server returned error, retrying...")
                    throw Exception("Failed to send summaries - server returned error")
                }
                
                Timber.i("📊 SUMMARY: Server accepted summaries")
                success = true
            }
            
            if (success) {
                Timber.i("📊 SUMMARY: Successfully sent ${summaries.size} summaries to server")
                CFResult.success(true)
            } else {
                Timber.w("📊 SUMMARY: Failed to send summaries after ${cfConfig.maxRetryAttempts} attempts")
                handleSendFailure(summaries)
                CFResult.error(
                    "Failed to send summaries after ${cfConfig.maxRetryAttempts} attempts",
                    category = ErrorHandler.ErrorCategory.NETWORK
                )
            }
        } catch (e: Exception) {
            Timber.e(e, "📊 SUMMARY: Error sending summaries to server")
            ErrorHandler.handleException(
                e,
                "Error sending summaries to server",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
            handleSendFailure(summaries)
            CFResult.error(
                "Error sending summaries to server: ${e.message}", 
                e, 
                category = ErrorHandler.ErrorCategory.NETWORK
            )
        }
    }
    
    /**
     * Helper method to handle failed summary send by re-queuing items
     */
    private fun handleSendFailure(summaries: List<CFConfigRequestSummary>) {
        Timber.w("Failed to send ${summaries.size} summaries after retries, re-queuing")
        var requeueFailCount = 0
        
        summaries.forEach { summary ->
            if (!this.summaries.offer(summary)) {
                requeueFailCount++
            }
        }
        
        if (requeueFailCount > 0) {
            ErrorHandler.handleError(
                "Failed to re-queue $requeueFailCount summaries after send failure",
                SOURCE,
                ErrorHandler.ErrorCategory.INTERNAL,
                ErrorHandler.ErrorSeverity.HIGH
            )
        }
    }

    /**
     * Starts the periodic flush timer
     */
    private fun startPeriodicFlush() {
        scope.launch {
            try {
                timerMutex.withLock {
                    // Cancel existing timer if any
                    flushTimer?.cancel()

                    // Create a new timer
                    flushTimer = fixedRateTimer(
                        "SummaryFlush",
                        daemon = true,
                        period = flushIntervalMs.get()
                    ) {
                        scope.launch {
                            try {
                                Timber.d("Periodic flush triggered for summaries")
                                flushSummaries()
                            } catch (e: Exception) {
                                ErrorHandler.handleException(
                                    e,
                                    "Error during periodic summary flush",
                                    SOURCE,
                                    ErrorHandler.ErrorSeverity.MEDIUM
                                )
                            }
                        }
                    }
                    Timber.d("Started periodic summary flush with interval ${flushIntervalMs.get()} ms")
                }
            } catch (e: Exception) {
                ErrorHandler.handleException(
                    e,
                    "Failed to start periodic summary flush",
                    SOURCE,
                    ErrorHandler.ErrorSeverity.HIGH
                )
            }
        }
    }

    /**
     * Restarts the periodic flush timer with the current interval
     */
    private suspend fun restartPeriodicFlush() {
        try {
            timerMutex.withLock {
                // Cancel existing timer if any
                flushTimer?.cancel()

                // Create a new timer with updated interval
                flushTimer = fixedRateTimer(
                    "SummaryFlush", 
                    daemon = true, 
                    period = flushIntervalMs.get()
                ) {
                    scope.launch {
                        try {
                            Timber.d("Periodic flush triggered for summaries")
                            flushSummaries()
                        } catch (e: Exception) {
                            ErrorHandler.handleException(
                                e,
                                "Error during periodic summary flush",
                                SOURCE,
                                ErrorHandler.ErrorSeverity.MEDIUM
                            )
                        }
                    }
                }
                Timber.d("Restarted periodic flush with interval ${flushIntervalMs.get()} ms")
            }
        } catch (e: Exception) {
            ErrorHandler.handleException(
                e,
                "Failed to restart periodic summary flush",
                SOURCE,
                ErrorHandler.ErrorSeverity.HIGH
            )
        }
    }

    /**
     * Returns all tracked summaries for other components
     * 
     * @return Map of experience IDs to tracking status
     */
    fun getSummaries(): Map<String, Boolean> = summaryTrackMap.toMap()
}

